### DISCUSSION TITLE
Migration from Gatsby to Next - polymorphic typescript issue

### DISCUSSION THREAD

**Discussion by davshoward** [vercel/next.js#45449] [Help] [answered]
I'm migrating a codebase from Gatsby to Next (largey pain free so far), but I've hit a snag I can't resolve and was hoping someone might be able to shed some light.

```
import type { LinkProps } from 'next/link';
import Link from 'next/link';
import React from 'react';

const Card = <TState,>({
	title,
	className,
	as: Component = 'article',
	...rest
}: {
	title: string;
	className?: string;
	as?: React.ElementType;
} & React.PropsWithoutRef<LinkProps<TState>>) => {
	return (
		<Component className="group rounded-2xl">
			<Link className="grid lg:grid-cols-12 gap-4 lg:gap-10 items-center" {...rest}>
				{title}
			</Link>
		</Component>
	);
};
```

But it throws the following error on the <Link>: 

`Property 'href' is missing in type '{ children: ("" | Element | undefined)[]; className: string; }' but required in type 'InternalLinkProps'.`

All I did was swap out GatsbyLinkProps for LinkProps.

Does anyone have idea where I'm going wrong?

**Comment by icyJoseph** [comment, marked as answer]
I guess the problem is that `LinkProps` doesn't take any generic parameters itself. 
Try with something like:

```tsx
import type { LinkProps } from "next/link";
import Link from "next/link";
import React, { ElementType } from "react";

const Card = <TState,>({
  title,
  className,
  as: Component = "article",
  linkAs,
  ...rest
}: {
  title: string;
  className?: string;
  as?: ElementType;
} & React.PropsWithoutRef<
  TState & Omit<LinkProps, "as"> & { linkAs: LinkProps["as"] }
>) => {
  return (
    <Component className="group rounded-2xl">
      <Link
        className="grid lg:grid-cols-12 gap-4 lg:gap-10 items-center"
        {...rest}
        as={linkAs}
      >
        {title}
      </Link>
    </Component>
  );
};

```

## Edit

I noticed you named your prop `as`, and that'll be united with `LinkProps`'s `as`, and then components can be URLs too.... Watch out! 

> In this case I renamed it to `linkAs`

**Reply by davshoward** [reply]
You absolute gem! Thank you. 

I was totally blind to that approach.

**Reply by icyJoseph** [reply]
I made an edit, sorry, I missed that you had an `as` prop, somehow I read component only... that has implications. Pretty much you need to omit it.

I updated once again where I named the `as` that'd be forwarded to the link as, `linkAs`
